src/Stratis.Bitcoin/P2P/SelfEndpointTracker.cs

```diff
@@ -22,5 +33,48 @@ public bool IsSelf(IPEndPoint ipEndPoint)
         {
             return this.knownSelfEndpoints.Contains(ipEndPoint);
         }
+
+        /// <summary>Updates external IP address and peer score of the node.</summary>
+        public void UpdateAndAssignMyExternalAddress(IPEndPoint ipEndPoint, ConnectionManager connectionManager)
+        {
+            // if external IP address supplied this overrides all.
+            if (connectionManager.ConnectionSettings.ExternalIpFromNodeConfiguration != null)
+            {
+                connectionManager.ConnectionSettings.ExternalEndpoint = connectionManager.ConnectionSettings.ExternalIpFromNodeConfiguration;
+                return;
+            }
+
+            // If external IP address not supplied take first routeable bind address and set score to 10.
+            IPEndPoint nodeServerEndpoint = connectionManager.ConnectionSettings.Listen?.FirstOrDefault(x => x.Endpoint.Address.IsRoutable(false))?.Endpoint;
+            if (nodeServerEndpoint != null)
+            {
+                this.MyExternalAddress = new KeyValuePair<IPEndPoint, int>(nodeServerEndpoint, 10);
+                connectionManager.ConnectionSettings.ExternalEndpoint = this.MyExternalAddress.Key;
+                return;
+            }
+
+            // If none supplied or routableable take from version handshake.
+            if (ipEndPoint == null)
```

Comment:
Could this happen earlier in the method, so you can return out early ?  Or do you need to setup connectionManager.ConnectionSettings.ExternalEndpoint first ?